gdal/frmts/netcdf/netcdfsgwriterutil.cpp

```diff
@@ -0,0 +1,910 @@
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+#include "netcdfsgwriterutil.h"
+#include "netcdfdataset.h"
+
+namespace nccfdriver
+{
+    SGeometry_Feature::SGeometry_Feature(OGRFeature& ft)
+    {
+        this->hasInteriorRing = false;
+        OGRGeometry * defnGeometry = ft.GetGeometryRef();
+
+        if (defnGeometry == nullptr)
+        {
+            throw SGWriter_Exception_EmptyGeometry();
+        }
+
+        OGRGeometry& r_defnGeometry = *defnGeometry;
+
+        OGRwkbGeometryType ogwkt = defnGeometry->getGeometryType();
+        this->type = OGRtoRaw(ogwkt);
+
+
+        if (this->type == MULTIPOINT)
+        {
+            OGRMultiPoint& r_defnGeometryMP = dynamic_cast<OGRMultiPoint&>(r_defnGeometry);
+
+            // Set total node count
+            this->total_point_count = r_defnGeometryMP.getNumGeometries();
+
+            // The amount of nodes is also the amount of parts
+            for(size_t pc = 0; pc < total_point_count; pc++)
+            {
+                ppart_node_count.push_back(1);
+            }
+
+            // total part count ==  total node count
+            this->total_part_count = this->total_point_count;
+
+        }
+
+        else if (this->type == LINE)
+        {
+            OGRLineString& r_defnGeometryLine = dynamic_cast<OGRLineString&>(r_defnGeometry);
+            // to do: check for std::bad_cast somewhere?
+
+            // Get node count
+            this->total_point_count = r_defnGeometryLine.getNumPoints();
+
+            // Single line: 1 part count == node count
+            this->ppart_node_count.push_back(r_defnGeometryLine.getNumPoints());
+
+            // One part
+            this->total_part_count = 1;
+        }
+
+        else if(this->type == MULTILINE)
+        {
+            OGRMultiLineString& r_defnMLS = dynamic_cast<OGRMultiLineString&>(r_defnGeometry);
+            this->total_point_count = 0;
+            this->total_part_count = r_defnMLS.getNumGeometries();
+
+            // Take each geometry, just add up the corresponding parts
+
+            for(int itrLScount = 0; itrLScount < r_defnMLS.getNumGeometries(); itrLScount++)
+            {
+                OGRLineString & r_LS = dynamic_cast<OGRLineString&>(*r_defnMLS.getGeometryRef(itrLScount));
+                int pt_count = r_LS.getNumPoints();
+
+                this->ppart_node_count.push_back(pt_count);
+                this->total_point_count += pt_count;
+            }
+        }
+
+        else if(this->type == POLYGON)
+        {
+            OGRPolygon& r_defnPolygon = dynamic_cast<OGRPolygon&>(r_defnGeometry);
+
+            this->total_point_count = 0;
+            this->total_part_count = 0;
+
+            // Get node count
+            // First count exterior ring
+            OGRLinearRing & exterior_ring = *r_defnPolygon.getExteriorRing();
```

Comment:
Not sure if it would be good to be permissive to this, but I decide not (i.e. nullptr check before deref possibly throws an exception) unless you think that I should try to pass this through quietly somehow? [fixed]